File: python/Day9.py
def part_1() -> None:
    height_map: list[list[int]] = []
    lowest_truths: list[list[bool]] = []

    with open("../data/day9.txt") as dFile:
        for row in dFile.readlines():
            if (row := row.strip()) != "":
                height_map.append([])
                lowest_truths.append([])
                for num in row:
                    height_map[-1].append(int(num))
                    lowest_truths[-1].append(False)

    for r in range(len(height_map)):
        for c in range(len(height_map[r])):
            truths: list[bool] = []
            if r >= 1:
                truths.append(height_map[r][c] < height_map[r - 1][c])
            if r + 1 < len(height_map):
                truths.append(height_map[r][c] < height_map[r + 1][c])
            if c >= 1:
                truths.append(height_map[r][c] < height_map[r][c - 1])
            if c + 1 < len(height_map[r]):
                truths.append(height_map[r][c] < height_map[r][c + 1])

            lowest_truths[r][c] = False not in truths

    result: int = sum(
        height_map[r][c] + 1
        for r in range(len(height_map))
        for c in range(len(height_map[r]))
        if lowest_truths[r][c]
    )

    print("Day:  9 | Part: 1 | Result: ", result)


def backtrack_basin(
    height_map: list[list[int]],
    r: int,
    c: int,
    counter: int,
    visited: list[tuple[int, int]],
) -> int:
    if (r, c) in visited:
        return counter

    visited.append((r, c))

    if height_map[r][c] == 9:
        return counter

    counter += 1

    if r >= 1:
        counter = backtrack_basin(height_map, r - 1, c, counter, visited)
    if r + 1 < len(height_map):
        counter = backtrack_basin(height_map, r + 1, c, counter, visited)
    if c >= 1:
        counter = backtrack_basin(height_map, r, c - 1, counter, visited)
    if c + 1 < len(height_map[r]):
        counter = backtrack_basin(height_map, r, c + 1, counter, visited)

    return counter


def part_2() -> None:
    height_map: list[list[int]] = []
    lowest_truths: list[list[bool]] = []

    with open("../data/day9.txt") as dFile:
        for row in dFile.readlines():
            if (row := row.strip()) != "":
                height_map.append([])
                lowest_truths.append([])
                for num in row:
                    height_map[-1].append(int(num))
                    lowest_truths[-1].append(False)

    for r in range(len(height_map)):
        for c in range(len(height_map[r])):
            truths: list[bool] = []
            if r >= 1:
                truths.append(height_map[r][c] < height_map[r - 1][c])
            if r + 1 < len(height_map):
                truths.append(height_map[r][c] < height_map[r + 1][c])
            if c >= 1:
                truths.append(height_map[r][c] < height_map[r][c - 1])
            if c + 1 < len(height_map[r]):
                truths.append(height_map[r][c] < height_map[r][c + 1])

            lowest_truths[r][c] = False not in truths

    basins: list[int] = []

    for r in range(len(height_map)):
        for c in range(len(height_map[r])):
            if lowest_truths[r][c]:
                basins.append(backtrack_basin(height_map, r, c, 0, []))

    basins.sort()
    print(basins)
    result: int = basins[-1] * basins[-2] * basins[-3]

    print("Day:  9 | Part: 2 | Result: ", result)


if __name__ == "__main__":
    part_1()
    part_2()

File: cs/Day9.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace cs
{
    class Day9
    {
        public static void Part1()
        {
            List<List<int>> height_map = new List<List<int>>();
            List<List<bool>> lowest_truths = new List<List<bool>>();

            File.ReadLines(@"../../../../../data/day9.txt").ToList().ForEach(row =>
            {
                if (row.Trim() != "")
                {
                    height_map.Add(new List<int>());
                    lowest_truths.Add(new List<bool>());

                    foreach (char num in row)
                    {
                        height_map.Last().Add(num - '0');
                        lowest_truths.Last().Add(false);
                    }
                }
            });

            for (int r = 0; r < height_map.Count; r++)
                for (int c = 0; c < height_map[r].Count; c++)
                {
                    List<bool> truths = new List<bool>();

                    if (r >= 1)
                        truths.Add(height_map[r][c] < height_map[r - 1][c]);
                    if (r + 1 < height_map.Count)
                        truths.Add(height_map[r][c] < height_map[r + 1][c]);
                    if (c >= 1)
                        truths.Add(height_map[r][c] < height_map[r][c - 1]);
                    if (c + 1 < height_map[r].Count)
                        truths.Add(height_map[r][c] < height_map[r][c + 1]);

                    lowest_truths[r][c] = truths.Contains(false) == false;
                }

            int result = 0;

            for (int r = 0; r < height_map.Count; r++)
                for (int c = 0; c < height_map[r].Count; c++)
                    if (lowest_truths[r][c])
                        result += height_map[r][c] + 1;

            Console.WriteLine("Day:  9 | Part: 1 | Result: " + result);
        }

        private static int BacktrackBasin(ref List<List<int>> height_map, int r, int c, int counter, ref List<Tuple<int, int>> visited)
        {
            if (visited.Contains(new Tuple<int, int>(r, c)))
                return counter;

            visited.Add(new Tuple<int, int>(r, c));

            if (height_map[r][c] == 9)
                return counter;

            counter++;

            if (r >= 1)
                counter = BacktrackBasin(ref height_map, r - 1, c, counter, ref visited);
            if (r + 1 < height_map.Count)
                counter = BacktrackBasin(ref height_map, r + 1, c, counter, ref visited);
            if (c >= 1)
                counter = BacktrackBasin(ref height_map, r, c - 1, counter, ref visited);
            if (c + 1 < height_map[r].Count)
                counter = BacktrackBasin(ref height_map, r, c + 1, counter, ref visited);

            return counter;
        }

        public static void Part2()
        {
            List<List<int>> height_map = new List<List<int>>();
            List<List<bool>> lowest_truths = new List<List<bool>>();

            File.ReadLines(@"../../../../../data/day9.txt").ToList().ForEach(row =>
            {
                if (row.Trim() != "")
                {
                    height_map.Add(new List<int>());
                    lowest_truths.Add(new List<bool>());

                    foreach (char num in row)
                    {
                        height_map.Last().Add(num - '0');
                        lowest_truths.Last().Add(false);
                    }
                }
            });

            for (int r = 0; r < height_map.Count; r++)
                for (int c = 0; c < height_map[r].Count; c++)
                {
                    List<bool> truths = new List<bool>();

                    if (r >= 1)
                        truths.Add(height_map[r][c] < height_map[r - 1][c]);
                    if (r + 1 < height_map.Count)
                        truths.Add(height_map[r][c] < height_map[r + 1][c]);
                    if (c >= 1)
                        truths.Add(height_map[r][c] < height_map[r][c - 1]);
                    if (c + 1 < height_map[r].Count)
                        truths.Add(height_map[r][c] < height_map[r][c + 1]);

                    lowest_truths[r][c] = truths.Contains(false) == false;
                }

            List<int> basins = new List<int>();
            List<Tuple<int, int>> visited = new List<Tuple<int, int>>();  

            for (int r = 0; r < height_map.Count; r++)
                for (int c = 0; c < height_map[r].Count; c++)
                    if (lowest_truths[r][c])
                        basins.Add(BacktrackBasin(ref height_map, r, c, 0, ref visited));

            basins.Sort();
            int result = basins[^1] * basins[^2] * basins[^3];

            Console.WriteLine("Day:  9 | Part: 1 | Result: " + result);
        }
    }
}
